example/conditionals/README.md

```diff
@@ -0,0 +1,49 @@
+# Conditional Rendering of Individual Resources
+
+In this example, we show how each resource in a pipeline step can
+be individually rendered based on a condition. We will have a blue and a
+green deployment that the user can activate:
+
+```yaml
+apiVersion: nop.example.org/v1alpha1
+kind: XNopConditional
+metadata:
+  name: test-resource
+spec:
+  env: dev
+  render: true
+  deployment:
+    blue: true
+    green: false
+```
+
+In our Composition both the `blue-resource` and the `green-resource` have a
+`condition` that determines if they will be run.
+
+```yaml
+    resources:
+         - name: blue-resource
+           condition: observed.composite.resource.spec.deployment.blue == true
```

Comment:
Could this match the existing P&T UX more closely?

Here we're "leaking" the structure of the `RunFunctionRequest`. I like the way it works and reads in isolation, but it's quite different from the rest of P&T where instead you'll see (e.g.) `FromCompositeFieldPath` and then a field path rooted at the composite.

The P&T API also doesn't expose `observed` vs `desired` anywhere else. In practice all "reads" (i.e. patches from) are always from observed state.

If we're confident that expressions will always be derived from the observed XR it could just be:

```yaml
resources:
- name: blue-resource
  condition: spec.deployment.blue == true  # This is always using the observed XR.
```

I'm guessing folks might want to use the environment, or other composed resources in their expressions though. So maybe:

```yaml
resources:
- name: blue-resource
  condition:
    type: CompositeFieldPath
    expression: spec.deployment.blue == true

- name: red-resource
  condition:
    type: EnvironmentFieldPath
    expression: data.red == true

- name: purple-resource
  condition:
    type: ComposedFieldPath
    composedResourceName: blue-resource
    expression: status.purple == true

```

Reply:
As I understand the suggestion will limit conditions to a single input. I can see cases where more complex conditions are built from a combination of sources, such as multiple composed/observed resources

```
- name: purple-resource
  condition:
    type: ComposedFieldPath
    composedResourceNames: 
       - blue-resource
       - red-resource
    expression: blue-resource.status.purple && red-resource.status.red
 ```
 
 This could also extend to needing to support referencing environment etc in the same condition.

Reply:
Yeah I don't like my suggestion very much. It does seem like using plain CEL would be more expressive and flexible. It's just a little out of place.

Maybe we should just drop the desired/observed prefix and always use observed state?

Reply:
While the nomenclature is a little odd for P&T, it matches function-go-templating and other functions in accessing the fields of the request: 

- function-go-templating <https://github.com/crossplane-contrib/function-go-templating?tab=readme-ov-file#using-this-function>
- KCL uses `option("params").oxr.spec.provider` <https://github.com/crossplane-contrib/function-kcl/blob/3b905d158bec3123d9e33f4801ba21d33ba685c2/examples/resources/condition/composition.yaml#L24>
- CUE func exposes the request param <https://github.com/crossplane-contrib/function-kcl/tree/main/examples/resources>

Reply:
Yeah. I still don't feel great about it though. 😄 

It's fair to say the pattern is idiomatic for functions in general. This is because the pattern "leaks" the internal shape of the `RunFunctionRequest`. I think that's fine - `observed.composed.resource` reads pretty well to me.

The challenge is that we don't leak that shape anywhere else in _this_ function because its API predates functions.

Reply:
Here are my two cents: One of the fundamental reasons why the functions were introduced was that classic patching was not expressive enough to cover many common scenarios (including conditional resources). 

Why I see benefits of having something consistent with patching (like the examples proposed by @negz), I do not find them as easy to read or as expressive as the one-liners proposed in the original PR. 

Typical Crossplane composition is a wall of text anyway. Simplification of expressions in conditions is a good way to prevent further text inflation. CEL has the benefit of being very expressive, easily understandable, and very efficient in terms of information-to-characters ratio. As everything will be much more complex than typical examples, I would avoid unnecessary wrappers (like the scenario presented by @delautour ) 

```
- name: purple-resource
  condition:
    type: ComposedFieldPath
    **composedResourceNames: 
       - blue-resource
       - red-resource**
    expression: blue-resource.status.purple && red-resource.status.red
```

The bolded part does not bring any value per se, but it should probably be included as combined conditions should mimic combined patches if consistency is the main criterion. Which will only decrease the readability of already convoluted compositions.

After all Occam's razor is always the best principle to follow when in doubt.

Reply:
I think we've discussed this before, but another option that wouldn't feel as much like it was mixing expression styles in one function would be to make conditionals a separate function. For example:

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: example
spec:
  # Omitted for brevity.
  mode: Pipeline
  pipeline:

  - step: patch-and-transform
    functionRef:
      name: function-patch-and-transform
    input:
      apiVersion: pt.fn.crossplane.io/v1beta1
      kind: Resources
      resources:
      - name: bucket
        base:
          apiVersion: s3.aws.upbound.io/v1beta1
          kind: Bucket
          spec:
            forProvider:
              region: us-east-2
        patches:
        - type: FromCompositeFieldPath
          fromFieldPath: "spec.location"
          toFieldPath: "spec.forProvider.region"
          transforms:
          - type: map
            map: 
              EU: "eu-north-1"
              US: "us-east-2"

  # This function filters resources produced by previous functions in the pipeline.
  # It matches using resource name. Any resource without a filter is not filtered. 
  - step: conditional-filters
    functionRef:
      name: function-cel-filter
    input:
      apiVersion: cel.fn.crossplane.io/v1beta1
      kind: Filters
      filters:
      - name: bucket
        condition: observed.composite.resource.spec.env == "prod" && observed.composite.resource.spec.render == true
```

Pros:

* You could use this with any function, not just P&T.
* P&T keeps a single expression style, not P&T&CEL.

Cons:

* Conditionals aren't defined near where resources are defined.

Reply:
I like this approach. It effectively pushes the approach that functions are highly specialized blocks and there is clean separation of responsibility. But the downside is that we will have to render resources and all their patches and then filter them out. This will bring performance hit for larger compositions with complex patching.

Also, it is a bit counter-intuitive from programming perspective to create everything and filter rather than just create conditionally.

Reply:
I went ahead and made this function: https://github.com/crossplane-contrib/function-cel-filter

Credit to @stevendborrelli, since it's pretty much his CEL code repackaged.

I'm leaning toward suggesting we try out using the separate filter function, rather than adding conditionals to function-patch-and-transform. We can always add them later if we find that folks really don't like using a separate function.